packages/dev/core/src/Misc/textureTools.ts

```diff
@@ -230,6 +231,179 @@ export function FromHalfFloat(value: number): number {
+                    }
+                }
+                if (!channels.G) {
+                    if (channels.R && !channels.B) {
+                        g = r;
+                    } else if (channels.B && !channels.R) {
+                        g = b;
+                    } else {
+                        g = 0;
+                    }
+                }
+                if (!channels.B) {
+                    if (channels.R && !channels.G) {
+                        b = r;
+                    } else if (channels.G && !channels.R) {
+                        b = g;
+                    } else {
+                        b = 0;
+                    }
+                }
+                data[i] = r;
+                data[i + 1] = g;
+                data[i + 2] = b;
+                data[i + 3] = a;
+            }
+        }
+
+        //To flip image on Y axis.
+        if ((texture as Texture).invertY || texture.isCube) {
+            for (let i = 0; i < halfHeight; i++) {
+                for (let j = 0; j < numberOfChannelsByLine; j++) {
+                    const currentCell = j + i * numberOfChannelsByLine;
+                    const targetLine = height - i - 1;
+                    const targetCell = j + targetLine * numberOfChannelsByLine;
+
+                    const temp = data[currentCell];
+                    data[currentCell] = data[targetCell];
+                    data[targetCell] = temp;
+                }
+            }
+        }
+
+        resolve(data);
+
+        // Unbind
+        engine.unBindFramebuffer(rtt.renderTarget);
+    } else {
+        reject();
+    }
+
+    rtt.dispose();
+    lodPostProcess.dispose();
+};
+
+/**
+ * Gets the data of the specified texture by rendering it to an intermediate RGBA texture and retreiving the bytes from it.
+ * This is convienent to get 8-bit RGBA values for a texture in a GPU compressed format.
+ * @param texture the source texture
+ * @param width the width of the result, which does not have to match the source texture width
+ * @param height the height of the result, which does not have to match the source texture height
+ * @param face if the texture has multiple faces, the face index to use for the source
+ * @param channels a filter for which of the RGBA channels to return in the result
+ * @param lod if the texture has multiple LODs, the lod index to use for the source
+ * @returns the 8-bit texture data
+ */
+export async function GetTextureDataAsync(
+    texture: BaseTexture,
+    width: number,
+    height: number,
+    face: number = 0,
+    channels: { R: boolean; G: boolean; B: boolean; A: boolean } = { R: true, G: true, B: true, A: true },
+    lod: number = 0
+): Promise<Uint8Array> {
+    return new Promise((resolve, reject) => {
+        if (!texture.isReady() && texture._texture) {
+            texture._texture.onLoadedObservable.addOnce(() => {
+                ProcessAsync(texture, width, height, face, channels, lod, resolve, reject);

```

Comment:
Not your code, but this is a dangling promise. If ProcessAsync fails, no one will be able to observe the exception.